be/src/olap/rowset/segment_v2/column_reader.cpp

```diff
@@ -0,0 +1,330 @@
+    if (page->offset_in_page == offset_in_page) {
+        // fast path, do nothing
+        return;
+    }
+
+    uint32_t pos_in_data = offset_in_page;
+    if (_reader->is_nullable()) {
+        rowid_t offset_in_data = 0;
+        rowid_t skips = offset_in_page;
+
+        if (offset_in_page > page->offset_in_page) {
+            // forward, reuse null bitmap
+            skips = offset_in_page - page->offset_in_page;
+            offset_in_data = page->data_decoder->current_index();
+        } else {
+            // rewind null bitmap, and
+            page->null_decoder = RleDecoder<bool>((const uint8_t*)page->null_bitmap.data, page->null_bitmap.size, 1);
+        }
+
+        auto skip_nulls = page->null_decoder.Skip(skips);
+        pos_in_data = offset_in_data + skips - skip_nulls;
+    }
+
+    page->data_decoder->seek_to_position_in_page(pos_in_data);
+    page->offset_in_page = offset_in_page;
+}
+
+Status FileColumnIterator::next_batch(size_t* n, ColumnVector* dst, MemPool* mem_pool) {
+    ColumnVectorView column_view(dst, mem_pool, _reader->type_info());
+    size_t remaining = *n;
+    while (remaining > 0) {
+        if (!_page->has_remaining()) {
+            bool eos = false;
+            RETURN_IF_ERROR(_load_next_page(&eos));
+            if (eos) {
+                break;
+            }
+        }
+
+        // number of rows to be read from this page
+        size_t nrows_in_page = std::min(remaining, _page->remaining());
+        size_t nrows_to_read = nrows_in_page;
+        if (_reader->is_nullable()) {
+            // when this column is nullable we read data in some runs
+            // first we read null bits in the same value, if this is null, we 
+            // don't need to read value from page.
+            // If this is not null, we read data from page in batch.
+            // This would be bad in case that data is arranged one by one, which
+            // will lead too many function calls to PageDecoder
+            while (nrows_to_read > 0) {
+                bool is_null = false;
+                size_t this_run = _page->null_decoder.GetNextRun(&is_null, nrows_to_read);
+
+                // we use num_rows only for CHECK
+                size_t num_rows = this_run;
+                if (!is_null) {
+                    RETURN_IF_ERROR(_page->data_decoder->next_batch(&num_rows, &column_view));
+                    DCHECK_EQ(this_run, num_rows);
+                }
+
+                // set null bits
+                column_view.set_null_bits(this_run, is_null);
+
+                nrows_to_read -= this_run;
+                _page->offset_in_page += this_run;
+                column_view.advance(this_run);
+                _current_rowid += this_run;
+            }
+        } else {
+            RETURN_IF_ERROR(_page->data_decoder->next_batch(&nrows_to_read, &column_view));
+            DCHECK_EQ(nrows_to_read, nrows_in_page);
+
+            if (column_view.is_nullable()) {
```

Comment:
if the column reader is not nullable, then will column_view.is_nullable can be true?